### ISSUE TITLE
DeferredFilesystemLock should return a cancellable Deferred

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6720]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#6720|
|Type|enhancement|
|Created|2013-08-31 21:16:21Z|
|Branch|https://github.com/twisted/twisted/tree/deferredfilesystemlock-cancellation-6720|

It already implements its own timeout internally; it would be nice if you could trigger this in advance. (Maybe this could even be refactored to use generic timeout support!)

<details><summary>Searchable metadata</summary>

```
trac-id__6720 6720
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_deferredfilesystemlock_cancellation_6720 branches/deferredfilesystemlock-cancellation-6720
branch_author__kaizhang kaizhang
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1377983781000000 1377983781000000
changetime__1383398982000000 1383398982000000
version__None None
owner__kaizhang kaizhang

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|kaizhang commented|
|-|-|

(In [39907]) Branching to 'deferredfilesystemlock-cancellation-6720'

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|kaizhang removed owner<br>kaizhang set status to `new`|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deferredfilesystemlock-cancellation-6720)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_4"></a>|@tomprince set owner to kaizhang|
|-|-|

Thanks. One minor point of grammar: "When cancelling" not "When cancel". Please merge after fixing that.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set status to `closed`|
|-|-|

(In [40558]) Merge deferredfilesystemlock-cancellation-6720: DeferredFilesystemLock supports Deferred cancellation.

Author: kaizhang
Reviewer: tom.prince
Fixes: [#6720](https://github.com/twisted/twisted/issues/6720)

The Deferred returned by twisted.internet.defer.DeferredFilesystemLock.deferUntilLocked can now be cancelled.
